### ISSUE TITLE
Test counter does not work

### ISSUE THREAD

**Issue by renatoathaydes** [fukamachi/rove#60]
There seems to be several issues with the test count being reported.

I have a few `deftest`s, and within each, lots of `testing` assertions.

The report shows all `testing` assertions, which is great, but they do not seem to "count".

But even `deftest` does not appear to "count"... not even the number of test files seems to be reflected in "count".

[Here](https://github.com/renatoathaydes/lisp-site-gen/actions/runs/5365715806/jobs/9734707007) you can see an example of my project's test run on GitHub Actions.

No matter how many tests I have , it reports at the end:

```
✓ 1 test completed

Summary:
  All 1 test passed.
```

When there are test failures (more than one), I still get count=1:

```
× 1 of 1 test failed

0) lisp-site-gen.parser-tests
     › PARSE-MARKDOWN-TEXT
       › NIL
   Expect (EQUAL (REVERSE LISP-SITE-GEN.PARSER-TESTS::OUTPUT)
                 `((:P "hi")
                   (:PRE
                    (:CODE :LANG "lisp"
                     ,(FORMAT NIL "(def main () (hello))~%(done)"))))) to be true.
     (EQUAL #1=(REVERSE LISP-SITE-GEN.PARSER-TESTS::OUTPUT)
            #2=`(#3=(:P "hi")
                 (:PRE
                  (:CODE :LANG #4="lisp"
                   ,(FORMAT NIL "(def main () (hello))~%(done)")))))
         OUTPUT = (#5=(:CODE-BLOCK "(def main () (hello))
     (done)"
                       "lisp")
                   #6=(:P "hi"))
         #1# = (#6# #5#)
         #2# = (#3#
                (:PRE
                 (:CODE :LANG #4# "(def main () (hello))
     (done)")))

1) lisp-site-gen.integration-tests
     › INTEGRATION-TEST
       › NIL
   Expect (EQUAL
           (WITH-OUTPUT-TO-STRING (LISP-SITE-GEN.INTEGRATION-TESTS::STR)
             (LISP-SITE-GEN:RUN '("tests/example.md")
                                LISP-SITE-GEN.INTEGRATION-TESTS::STR)
             LISP-SITE-GEN.INTEGRATION-TESTS::STR)
           (CONCATENATE 'STRING "<H1>Header</H1>"
                        "<P></P><P>This is a paragraph.</P><P></P>"
                        (FORMAT NIL
                                "<QUOTATION> Quotes may~%contain multi-line.</QUOTATION>")
                        "<H2>Something.</H2>")) to be true.
     (EQUAL
      #1=(WITH-OUTPUT-TO-STRING (LISP-SITE-GEN.INTEGRATION-TESTS::STR)
           (LISP-SITE-GEN:RUN '("tests/example.md")
                              LISP-SITE-GEN.INTEGRATION-TESTS::STR)
           LISP-SITE-GEN.INTEGRATION-TESTS::STR)
      #2=(CONCATENATE 'STRING "<H1>Header</H1>"
                      "<P></P><P>This is a paragraph.</P><P></P>"
                      #3=(FORMAT NIL
                                 "<QUOTATION> Quotes may~%contain multi-line.</QUOTATION>")
                      "<H2>Something.</H2>"))
         #1# = "<H1>Header</H1><P></P><P>This is a paragraph.</P><P></P><QUOTATION> Quotes may
     contain multi-line.</QUOTATION><H2>Something.</H2><P></P><P><SPAN>This is <CODE>Lisp</CODE> code:</SPAN></P><P></P><CODE-BLOCK>(defun looks-cool
         (it-does)) lisp</CODE-BLOCK>"
         #3# = "<QUOTATION> Quotes may
     contain multi-line.</QUOTATION>"
         #2# = "<H1>Header</H1><P></P><P>This is a paragraph.</P><P></P><QUOTATION> Quotes may
     contain multi-line.</QUOTATION><H2>Something.</H2>"

Summary:
  1 test failed.
    - lisp-site-gen/tests
```

Did I configure something wrong? Why is the counter always one for everything?

**Comment by kevpand** [comment]
Getting this as well, when I run `rove my-project.asd` on my project in a Github action, I always get all tests passed with a test count of 1, no matter how many tests and if they're are failures or not.

**Comment by renatoathaydes** [comment]
I ended up just using fiveam. It's good enough for my needs and it works.

**Comment by fukamachi** [comment]
@kevpand
Can I see your project Actions? (if it's public)
Will you mind trying the `rove` execution on your machine and show me the result of `echo $?`?

**Comment by kevpand** [comment]
Sure! The local output is attached, the project actions isn't public but it's very close to [this roswell example](https://roswell.github.io/GitHub-Actions.html)

I run the code locally against my asd file and see the same, I have two tests written but only see that one test was run and the output stays the same. If I run `asdf:test-system 'my-system` in my REPL however both tests run.

**Comment by fukamachi** [comment]
It may depend on the details of how the test is written and organized.

I made [an example project](https://github.com/fukamachi/rove-test-example) to run CI with Rove and GitHub Actions.
The test fails intentionally, and it seems to work.

https://github.com/fukamachi/rove-test-example/actions/runs/6079330345/job/16491671093

Can you find a different point from yours?

**Comment by kevpand** [comment]
Yeah I think it is, my asd file looks like the following:

```(in-package #:asdf-user)

(defsystem "project"
  :class :package-inferred-system
  :version "0.1.0"
  :author "Kevin Anderson"
  :license "LLGPL v2.1"
  :depends-on ("alexandria"
               "project/package")
  :description "Tool to help manage ZOIA patches"
  :in-order-to ((test-op (test-op "project/tests"))))

(defsystem "project/tests"
  :class :package-inferred-system
  :author "Kevin Anderson"
  :license "LLGPL v2.1"
  :depends-on ("rove"
               "trivial-gray-streams"
               "project/tests/package")
  :description "Test system for alex"
  :perform (test-op (op c) (symbol-call
                            :alex/tests/package
                            :run-all-tests)))
```

and my `tests/package.lisp`:

```
(uiop:define-package #:project/tests/package
  (:nicknames #:project/tests)
  (:use #:cl)
  (:import-from #:rove)
  (:import-from #:project/tests/debug))

(in-package #:project/tests)

(defparameter *test-suites*
  '(:project/tests/debug))

(defun run-all-tests ()
  (dolist (suite *test-suites*)
    (rove:run-suite suite)))
```

**Comment by fukamachi** [comment]
Ah, I see the reason.

`rove:run-suite` is intended to be run internally or in REPL, so it doesn't record the results to `*last-suite-report*`, which is used to detect if the tests passed or not.

**Comment by fukamachi** [comment]
Perhaps it will work if you rewrite it like this:

```common-lisp
(defun run-all-tests ()
  (rove:run-test-functions
    (list (lambda ()
            (dolist (suite *test-suites*)
              (rove:run-suite suite))))))
```

**Comment by kevpand** [comment]
Had to make a slight modification but that works!

```
(defun run-all-tests ()
  (rove/core/suite::run-test-functions
    (list (lambda ()
            (dolist (suite *test-suites*)
              (rove:run-suite suite))))))
```
Doesn't appear that `run-test-functions` is exported from the `rove` package namespace :/

Would it be preferable to just list the tests in the asd file explicitly rather than the package file? I did that so the package and test asd systems would be symmetric.

**Comment by fukamachi** [comment]
The `package-inferred-system` is fairly new to the Common Lisp world, and I'm still unsure how to organize tests with it.

In my daytime project, I added `./tests.lisp` like:

```common-lisp
(defpackage #:myproject/tests
  (:use #:cl)
  (:import-from #:myproject/tests/test1)
  (:import-from #:myproject/tests/test2)
  (:import-from #:myproject/tests/test3))
```

Then, you can run all test suites with `rove:run` like:

```common-lisp
(rove:run :myproject/tests)
```

Anyway, you need to list all test suites somewhere, in a ASD file, in a defpackage, or in a list and run them by `run-suite`.

**Comment by kevpand** [comment]
Oh yeah no worries! Thanks for your help really appreciate all you do!

**Comment by fukamachi** [comment]
I'm closing this since it has been fixed by #64. Sorry for the delay.
